src/openms_gui/source/VISUAL/TOPPASScene.cpp

```diff
@@ -480,26 +488,64 @@ namespace OpenMS
 
   bool TOPPASScene::isEdgeAllowed_(TOPPASVertex* u, TOPPASVertex* v)
   {
-    if (u == 0 ||
-        v == 0 ||
-        u == v ||
+    if (u == 0 || v == 0 || u == v ||
         // edges leading to input files make no sense:
         qobject_cast<TOPPASInputFileListVertex*>(v) ||
         // neither do edges coming from output files:
         qobject_cast<TOPPASOutputFileListVertex*>(u) ||
-        // nor edges from input to output without a tool in between:
-        (qobject_cast<TOPPASInputFileListVertex*>(u)
-        && qobject_cast<TOPPASOutputFileListVertex*>(v)) ||
-        // nor multiple incoming edges for a single output file/list node
-        (qobject_cast<TOPPASOutputFileListVertex*>(v)
-        && v->inEdgesBegin() != v->inEdgesEnd()) ||
-        // nor mergers connected directly to an output node
-        (qobject_cast<TOPPASMergerVertex*>(u)
-        && qobject_cast<TOPPASOutputFileListVertex*>(v)))
+        // nor edges from input/merger/splitter directly to output:
+        ((qobject_cast<TOPPASInputFileListVertex*>(u) || qobject_cast<TOPPASMergerVertex*>(u) || qobject_cast<TOPPASSplitterVertex*>(u)) && qobject_cast<TOPPASOutputFileListVertex*>(v)) ||
+        // nor multiple incoming edges for an output or splitter node:
+        ((qobject_cast<TOPPASOutputFileListVertex*>(v) || qobject_cast<TOPPASSplitterVertex*>(v)) && (v->inEdgesBegin() != v->inEdgesEnd())))
     {
       return false;
     }
 
+    // can't have more incoming edges than a tool has inputs:
+    TOPPASToolVertex* tv = qobject_cast<TOPPASToolVertex*>(v);
+    if (tv)
+    {
+      QVector<TOPPASToolVertex::IOInfo> input_infos;
+      tv->getInputParameters(input_infos);
+      if (tv->incomingEdgesCount() >= Size(input_infos.size())) return false;
+      // also, no edges from collectors to tools without input file lists:
+      // @TODO: what if the input file list is already occupied by an edge?
+      TOPPASMergerVertex* mv = qobject_cast<TOPPASMergerVertex*>(u);
+      if (mv && !mv->roundBasedMode())
```

Comment:
Well, once it's for input and once for output. If I add functions, I think it would make most sense to have them in TOPPASToolVertex, as `hasInputList` and `hasOutputList`.